echopype/convert/parse_base.py

```diff
@@ -69,54 +76,283 @@ def _print_status(self):
+                    use_swap=use_swap,
+                    zarr_root=zarr_root,
+                    max_chunk_size=max_chunk_size,
+                )
 
-    def rectangularize_transmit_ping_data(self, data_type: str) -> None:
-        """
-        Rectangularize the ``data_type`` data within transmit ping data.
-        Additionally, convert the data to a numpy array
-        indexed by channel.
+    @staticmethod
+    def _write_to_temp_zarr(
+        arr: np.ndarray,
+        zarr_root: zarr.Group,
+        path: str,
+        shape: Tuple[int],
+        chunks: Tuple[int],
+    ) -> dask.array.Array:
+        if shape == arr.shape:
+            z_arr = zarr_root.array(
+                name=path,
+                data=arr,
+                fill_value=np.nan,
+                chunks=chunks,
+                dtype="f8",
+                write_empty_chunks=False,
+            )
+        else:
+            # Figure out current data region
+            region = tuple([slice(0, i) for i in arr.shape])
+
+            # Create zarr array
+            z_arr = zarr_root.full(
+                name=path,
+                shape=shape,
+                chunks=chunks,
+                dtype="f8",
+                fill_value=np.nan,  # same as float64
+                write_empty_chunks=False,
+            )
 
-        Parameters
-        ----------
-        data_type: str
-            The key of ``self.ping_data_dict_tx`` to rectangularize
-        """
+            # Fill zarr array with actual data
+            z_arr.set_basic_selection(region, arr)
+
+        # Set dask array from zarr array
+        d_arr = da.from_zarr(z_arr)
+        return d_arr
+
+    def _parse_and_pad_datagram(
+        self,
+        data_type,
+        data_type_shapes: dict = {},
+        raw_type: Literal["transmit", "receive"] = "receive",
+        use_swap: bool = False,
+        zarr_root: Optional[zarr.Group] = None,
+        max_chunk_size: str = "100MB",
+    ) -> None:
+        ping_data_dict = self.ping_data_dict_tx if raw_type == "transmit" else self.ping_data_dict
+
+        # If there's no data, set and skip
+        if data_type_shapes[data_type] is None:
+            no_data_dict = {ch_id: None for ch_id in ping_data_dict[data_type].keys()}
+            ping_data_dict[data_type] = no_data_dict
+            return
+
+        # Set up zarr when using swap
+        # by determining the chunk sizes
+        if use_swap:
+            if zarr_root is None:
+                raise ValueError("zarr_root cannot be None when use_swap is True")
+
+            # Get the final data shape
+            data_shape = data_type_shapes[data_type]
+
+            # Auto chunk on first dimension
+            # since this is ping time
+            chunks = ("auto",) + (data_shape[1],)
+            chunks = auto_chunks(
+                chunks=chunks, shape=data_shape, limit=max_chunk_size, dtype=np.dtype("float64")
+            )
+            chunks = chunks + data_shape[2:]
```

Comment:
Adding some explanation -- is this correct?
```suggestion
            chunks = chunks + data_shape[2:]  # angle data has >2 dimensions
```

Reply:
I think this part does not need to change if complex data is reshaped in the parser and `data_shape[2:]` in that case would contain the right shape, i.e.:
- angle data: (number_of_pings, length_of_range_sample_dim, 2)
- complex data: (number_of_pings, length_of_range_sample_dim, n_sector), where `n_sector` is 3 or 4

Reply:
The comment is almost correct, it can also cover complex data like you mentioned. The code snippet above determines the chunk sizes for each dimension:

- L248: Create chunks variable like ("auto", length_of_range_sample_dim)
- L249: Dask `auto_chunks` determine the most optimized chunk size for the "auto" dimension, in this case ping time
- L252: This one would take chunks determined by dask and we append to the rest of the dimension sizes, since angle and complex might have more that 2D, then this would be captured and the resulting chunks might look something like `((100,), (5000,), 2)`. This why we have L253 that grabs the first value of chunks to create something like `(100, 5000, 2)`. 
 
> I think this part does not need to change if complex data is reshaped in the parser and data_shape[2:] in that case would contain the right shape,

Yea, you're right in this comment that no matter what the rest of the dimension is, it will just keep the chunk as the full dimension sizes.